src/lens.jl

```diff
@@ -146,22 +135,48 @@ function get(l::ComposedLens, obj)
     get(l.lens1, inner_obj)
 end
 
-function set(l::ComposedLens, obj, val, m::MutationPolicy)
+function set(l::ComposedLens, obj, val)
     inner_obj = get(l.lens2, obj)
-    inner_val = set(l.lens1, inner_obj, val, m)
-    set(l.lens2, obj, inner_val, m)
+    inner_val = set(l.lens1, inner_obj, val)
+    set(l.lens2, obj, inner_val)
 end
 
 struct IndexLens{I <: Tuple} <: Lens
     indices::I
 end
 
 get(l::IndexLens, obj) = getindex(obj, l.indices...)
-set(l::IndexLens, obj, val, ::ForbidMutation) = Base.setindex(obj, val, l.indices...)
-function set(l::IndexLens, obj, val, ::EncourageMutation)
-    if applicable(setindex!, obj, val, l.indices)
-        setindex!(obj, val, l.indices...)
-    else
-        set(l, obj, val, ForbidMutation())
+set(l::IndexLens, obj, val) = setindex(obj, val, l.indices...)
+
+
+const NNamedTupleLens{N,s} = NamedTuple{s, T} where {T <: NTuple{N, Lens}}
+struct MultiPropertyLens{L <: NNamedTupleLens} <: Lens
+    lenses::L
```

Comment:
Why not use just a tuple of lenses, rather than a named tuple?  With a named tuple, you can only support properties, right?  (Though maybe that was the intention as you named it Multi**Property**Lens?)  I mean, I wonder if it makes sense to support "vectorizing" lens of mixed kinds, like this:

```julia
obj0 = (a=1, b=2, c=3)
l = @lens(  # this is a made-up syntax
  _.a,      # get/set by property
  _[3],     # get/set by index
)
obj1 = set(l, obj0, (100, 300))
@assert obj1 == (a=100, b=2, c=300)
```

Reply:
Motivation was that I had lots of code, which looks like
```julia
obj = @set obj.a = a1
obj = @set obj.b = b1
obj = @set obj.c = c1
```
And the compiler was often unable to make this as fast as some handwritten
```julia
obj = Obj(a1,b1,c1, obj.other_fields...)
```
I like your approach, but I think it is harder for the compiler. I need to play around with it. Anyway this is a separate concern from removing mutation policy. It was a mistake to add it to this branch and I will remove it.

Reply:
Yeah, I agree that it's nice to release a simple version now and do some design iteration.  For the multi-lens, let's continue discussion and design iteration in https://github.com/jw3126/Setfield.jl/issues/23#issuecomment-413039326